ocsp.go

```diff
@@ -407,101 +515,212 @@ func verifyPeerCertificate(callback func(*sync.WaitGroup, []*x509.Certificate) [
 			verifiedChains[i] = append(verifiedChains[i], rca)
 			n++
 		}
-		wg.Add(n)
-		results := callback(&wg, verifiedChains[i])
-		wg.Wait()
+		results := getAllRevocationStatus(verifiedChains[i])
 		for _, r := range results {
-			if r.err != nil {
+			if r.err != nil && (r.code == ocspRevoked || ocspFailOpen == ocspFailOpenFalse) {
+				// return error if revoked or OCSPFailOpen mode is false
 				return r.err
 			}
 		}
 	}
-	writeOCSPCacheFile()
+
+	ocspResponseCacheLock.Lock()
+	if cacheUpdated {
+		writeOCSPCacheFile()
+	}
+	cacheUpdated = false
+	ocspResponseCacheLock.Unlock()
 	return nil
 }
 
-func getAllRevocationStatusParallel(wg *sync.WaitGroup, verifiedChains []*x509.Certificate) []*ocspStatus {
+func validateWithCacheForAllCertificates(verifiedChains []*x509.Certificate) bool {
 	n := len(verifiedChains) - 1
-	ocspStatusChan := make(chan *ocspStatus, n)
 	for j := 0; j < n; j++ {
-		go getRevocationStatus(wg, ocspStatusChan, verifiedChains[j], verifiedChains[j+1])
+		subject := verifiedChains[j]
+		issuer := verifiedChains[j+1]
+		if valid, _, _ := validateWithCache(subject, issuer); !valid {
+			return false
+		}
 	}
-	results := make([]*ocspStatus, n)
-	for j := 0; j < n; j++ {
-		results[j] = <-ocspStatusChan // will wait for all results back
+	return true
+}
+
+func validateWithCache(subject, issuer *x509.Certificate) (bool, []byte, *certIDKey) {
+	ocspReq, err := ocsp.CreateRequest(subject, issuer, &ocsp.RequestOptions{})
+	if err != nil {
+		glog.V(2).Infof("failed to create OCSP request from the certificates.\n")
+		return false, nil, nil
 	}
-	close(ocspStatusChan)
-	return results
+	encodedCertID, ocspS := extractCertIDKeyFromRequest(ocspReq)
+	if ocspS.code != ocspSuccess {
```

Comment:
nit: using ocspSuccess as error codes can be a misnomer and confuse someone about the functionality of the function. This function does not process ocsp status and hence should not ideally have return value named ocspSuccess. More generic error code names like "success" "failure" would be better.

Reply:
Agree. Separating function is success and revocation status good cases.